tests/component/_digital_utils.py

```diff
@@ -171,14 +171,14 @@ def _create_waveform_for_line(num_samples: int, line_number: int) -> DigitalWave
 
 def _create_waveform_for_lines(
     num_samples: int, first_line: int, num_lines: int, dtype: type[numpy.generic]
-) -> DigitalWaveform:
+) -> DigitalWaveform[Any]:
```

Comment:
Functions that take a dtype (this and _create_digital_waveform) should use a typevar to propagate it to the return type.
```python
def _create_waveform_for_lines(
    num_samples: int, first_line: int, num_lines: int, dtype: type[_D]
) -> DigitalWaveform[_D]:
```

(If _D doesn't work, try nitypes.waveform.typing.TDigitalState.)